File: solutions/prolog/difference-of-squares/1/difference_of_squares.pl
sum_of_squares(0, 0).

sum_of_squares(N, Result) :-
    N > 0,
    Previous is N - 1,
    sum_of_squares(Previous, PreviousResult),
    Result is N * N + PreviousResult.

sum_to(0, 0).

sum_to(N, Result) :-
    N > 0,
    Previous is N - 1,
    sum_to(Previous, PreviousResult),
    Result is N + PreviousResult.

square_of_sum(N, Result) :-
    sum_to(N, Sum),
    Result is Sum * Sum.

difference(N, Result) :-
    square_of_sum(N, SquareSums),
    sum_of_squares(N, SumSquares),
    Result is SquareSums - SumSquares.

File: solutions/ocaml/difference-of-squares/1/difference_of_squares.ml

let rec sum_to number = 
  if number <= 0 then
    0
  else 
    number + sum_to (number-1)

let square_of_sum number = 
    let total = sum_to number in
    total * total

let rec sum_of_squares number =
    if number <= 0 then
      0
    else
      number*number + sum_of_squares (number-1)

let difference_of_squares number =
    square_of_sum number - sum_of_squares number
